test/puppetlabs/trapperkeeper/services_test.clj

```diff
@@ -49,56 +55,181 @@
+      (with-app-with-empty-config app
+        services
+        (is (= [:init-service1 :init-service2 :init-service3
+                :start-service1 :start-service2 :start-service3]
+               @call-seq))
+        (app/restart app)
+        (is (= [:init-service1 :init-service2 :init-service3
+                :start-service1 :start-service2 :start-service3
+                :stop-service3 :stop-service2 :stop-service1
+                :init-service1 :init-service2 :init-service3
+                :start-service1 :start-service2 :start-service3]
+               @call-seq)))
+      (is (= [:init-service1 :init-service2 :init-service3
+              :start-service1 :start-service2 :start-service3
+              :stop-service3 :stop-service2 :stop-service1
+              :init-service1 :init-service2 :init-service3
+              :start-service1 :start-service2 :start-service3
+              :stop-service3 :stop-service2 :stop-service1]
+             @call-seq)))))
+
+(deftest test-lifecycle-function-ordering-signaling
+  (testing "app restart calls life cycle functions in the correct order"
+    (let [call-seq (atom [])
+          services (create-lifecycle-services call-seq)]
+      (with-app-with-empty-config app
+        services
+        (is (= [:init-service1 :init-service2 :init-service3
+                :start-service1 :start-service2 :start-service3]
+               @call-seq))
+        (internal/register-sighup-handler [app])
+        (beckon/raise! "HUP")
+        (let [start (System/currentTimeMillis)]
+          (while (and (not= (count @call-seq) 15)
+                     (< (- (System/currentTimeMillis) start) 1000))
+            (Thread/yield)))
+        (is (= (count @call-seq) 15))
+        (is (= [:init-service1 :init-service2 :init-service3
+                :start-service1 :start-service2 :start-service3
+                :stop-service3 :stop-service2 :stop-service1
+                :init-service1 :init-service2 :init-service3
+                :start-service1 :start-service2 :start-service3]
+               @call-seq)))
+      (is (= [:init-service1 :init-service2 :init-service3
+              :start-service1 :start-service2 :start-service3
+              :stop-service3 :stop-service2 :stop-service1
+              :init-service1 :init-service2 :init-service3
+              :start-service1 :start-service2 :start-service3
+              :stop-service3 :stop-service2 :stop-service1]
+             @call-seq)))))
+
+(deftest test-context-cleared-on-restart
+  (testing "service contexts should be cleared out during a restart"
+    (let [test-context (atom {})
+          service1 (service EmptyService
+                     []
+                     (init [this context]
+                           (swap! test-context merge context)
+                           {:foo "bar"}))]
+      (with-app-with-empty-config app [service1]
+        (app/restart app))
+      (is (= {} @test-context)))))
```

Comment:
I think this might be easier to read if you just set the initial atom to `nil`, and used `reset!` instead of `swap!`/`merge`.

I also wouldn't mind adding a few more assertions in this test.  Like... maybe...

1. rename `test-context` to `test-init-context`
2. add a second atom that is called `test-start-context`
3. add a `start` lifecycle fn that has the same sort of `swap!` logic, to `reset!` its atom to the context that it was passed
4. Inside of the `with` block, add assertions for the value of the two atoms
5. still inside of the `with` block, reset the two atoms to `nil`
6. add an assertion on the second atom outside of the `with` block.

Bonus points for adding in one more atom that keeps a counter of the number of times `init` is called, and puts that counter into the context map that `init` returns, so that the assertions against the `start` atom can include the value of that counter.